### PR TITLE
fix: File lock issue git

### PR DESCRIPTION
**Pull request by AnaghHegde**
#### PR fixes following issue(s)
Fixes https://github.com/appsmithorg/appsmith/issues/24854

#### Type of change
- Bug fix (non-breaking change which fixes an issue)

#### How Has This Been Tested?
- [ ] Manual

#### Dev activity
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

**Review by nayan-rafiq — APPROVED** [review]

**Review by nayan-rafiq — APPROVED** [review]

File: app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java [hunk 1]
```diff
@@ -1067,4 +1069,27 @@ private JSONObject getMainContainer(Object pageJson, Gson gson) {
         JSONArray layouts = pageJSON.getJSONObject("unpublishedPage").getJSONArray("layouts");
         return layouts.getJSONObject(0).getJSONObject("dsl");
     }
+
+    @Override
+    public Mono<Long> deleteIndexLockFile(Path path, int validTime) {
```
Comment (nayan-rafiq) [inline app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java:1074]:
```suggestion
    public Mono<Long> deleteIndexLockFile(Path path, int validTimeInSeconds) {
```

File: app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java [hunk 2]
```diff
@@ -1067,4 +1069,27 @@ private JSONObject getMainContainer(Object pageJson, Gson gson) {
...
+        // Check the time created of the index.lock file
+        // If the File is stale for more than validTime, then delete the file
+        try {
+            BasicFileAttributes attr = Files.readAttributes(path, BasicFileAttributes.class);
+            FileTime fileTime = attr.creationTime();
+            Instant now = Instant.now();
+            Instant validCreateTime = now.minusSeconds(validTime);
```
Comment (nayan-rafiq) [inline app/server/appsmith-git/src/main/java/com/appsmith/git/helpers/FileUtilsImpl.java:1081]:
```suggestion
            Instant validCreateTime = now.minusSeconds(validTimeInSeconds);
```

File: app/server/appsmith-interfaces/src/main/java/com/appsmith/external/git/FileInterface.java [hunk 3]
```diff
@@ -73,4 +73,6 @@ Mono<ApplicationGitReference> reconstructApplicationReferenceFromGitRepo(
      * @return success if the clone repo doesnt contain any files
      */
     Mono<Boolean> checkIfDirectoryIsEmpty(Path baseRepoSuffix) throws IOException;
+
+    Mono<Long> deleteIndexLockFile(Path path, int validTime);
```
Comment (nayan-rafiq) [inline app/server/appsmith-interfaces/src/main/java/com/appsmith/external/git/FileInterface.java:77]:
```suggestion
    Mono<Long> deleteIndexLockFile(Path path, int validTimeInSeconds);
```

File: app/server/appsmith-server/src/main/java/com/appsmith/server/exceptions/GlobalExceptionHandler.java [hunk 4]
```diff
@@ -283,14 +305,77 @@ public Mono<ResponseDTO<ErrorDTO>> catchException(Exception e, ServerWebExchange
         return getResponseDTOMono(urlPath, response);
     }
 
+    @ExceptionHandler
+    @ResponseBody
+    public Mono<ResponseDTO<ErrorDTO>> catchJGitInternalException(JGitInternalException e, ServerWebExchange exchange) {
+        AppsmithError appsmithError = AppsmithError.GIT_ACTION_FAILED;
+        exchange.getResponse().setStatusCode(HttpStatus.resolve(appsmithError.getHttpErrorCode()));
+        doLog(e);
+        // lockFailedException
```
Comment (nayan-rafiq) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/exceptions/GlobalExceptionHandler.java:314]:
```suggestion
```

File: app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/GitFileUtils.java [hunk 5]
```diff
@@ -73,6 +74,9 @@ public class GitFileUtils {
 
     private final Gson gson;
 
+    @Value("${appsmith.index.lock.file.time}")
+    public static final int INDEX_LOCK_FILE_STALE_TIME = 900;
```
Comment (nayan-rafiq) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/helpers/GitFileUtils.java:78]:
```suggestion
    public static final int INDEX_LOCK_FILE_STALE_TIME = 900; // number of seconds after lock file is stale
```

**Review by nayan-rafiq — CHANGES_REQUESTED** [review]
I recommend to add some unit tests for this.

**Review by Parthvi12 — APPROVED** [review]

**Comment by AnaghHegde** [comment]
@Aishwarya-U-R Its not possible to add a server side test case for this scenario.

**Pull request merged**